homeassistant/components/combined_energy/sensor.py

```diff
@@ -0,0 +1,270 @@
+        sensor_factory.entities()
+    )
+    # Insert connected as the first sensor
+    entities.insert(0, CombinedEnergyConnectedSensor(entry.title, connection))
+
+    async_add_entities(entities)
+
+
+class CombinedEnergyConnectedSensor(CoordinatorEntity, BinarySensorEntity):
+    """Representation of a Combined Energy connection status sensor."""
+
+    data_service: CombinedEnergyConnectivityDataService
+
+    def __init__(
+        self, entry_title: str, data_service: CombinedEnergyConnectivityDataService
+    ) -> None:
+        """Initialise Connected Sensor."""
+        super().__init__(data_service.coordinator)
+
+        self.data_service = data_service
+        self.entity_description = SENSOR_DESCRIPTION_CONNECTED
+
+        self._attr_name = f"{entry_title} {self.entity_description.name}"
+        self._attr_unique_id = f"install_{self.data_service.api.installation_id}-{self.entity_description.key}"
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return the state of the sensor."""
+        if self.data_service.data is not None:
+            return self.data_service.data.connected
+        return None
+
+
+class CombinedEnergyReadingsSensor(CoordinatorEntity, SensorEntity):
+    """Representation of a Combined Energy API reading energy sensor."""
+
+    data_service: CombinedEnergyReadingsDataService
+    entity_description: SensorEntityDescription
+
+    native_value_rounding: int = 2
+
+    def __init__(
+        self,
+        device: Device,
+        device_info: DeviceInfo,
+        description: SensorEntityDescription,
+        data_service: CombinedEnergyReadingsDataService,
+    ) -> None:
+        """Initialise Readings Sensor."""
+        super().__init__(data_service.coordinator)
+
+        self.device_id = device.device_id
+        self.data_service = data_service
+        self.entity_description = description
+
+        self._attr_name = f"{device.display_name} {description.name}"
+        self._attr_device_info = device_info
+        self._attr_unique_id = (
+            f"install_{self.data_service.api.installation_id}-"
+            f"device_{device.device_id}-"
+            f"{description.key}"
+        )
+
+    @property
+    def device_readings(self) -> DeviceReadings | None:
+        """Get readings for specific device."""
+        if data := self.data_service.data:
+            return data.get(self.device_id, None)
+        return None
+
+    @property
+    def _raw_value(self):
+        """Get raw reading value from device readings."""
+        if device_readings := self.device_readings:
+            return getattr(device_readings, self.entity_description.key)
+        return None
+
+    @property
+    def available(self) -> bool:
+        """Indicate if the entity is available."""
+        return self._raw_value is not None
+
+    @abstractmethod
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
```

Comment:
instead of doing this, you can add a mixin to your sensor_entity descriptions of a value_fn that does what ever operations you need it to. There are plenty of examples in the code

Reply:
That approach doesn't work in this case. Each of the different sensor types handles the data from the API in different ways and accesses multiple member variables to do so. 
The Sensor types are selected based on the SensorDeviceClass.